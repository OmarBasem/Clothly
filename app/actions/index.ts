import {fetchUserAction} from "./user";
import {fetchProductsAction} from "./products";

export {fetchUserAction, fetchProductsAction}