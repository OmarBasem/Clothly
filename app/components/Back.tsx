import React from 'react';
import {TouchableOpacity, Platform, StyleSheet, GestureResponderEvent} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import {theme} from '../styles'

interface IBackProps {
    onPress: ((event: GestureResponderEvent) => void),
}

// Navigation back button
const Back = (props: IBackProps) => {
    return (
        <TouchableOpacity activeOpacity={1} style={s.back}
                          onPress={props.onPress} hitSlop={{right: 40}} testID='back'>
            <Icon name={Platform.OS === 'ios' ? 'chevron-back' : 'md-arrow-back'} size={40} color={theme.mainColor}/>
        </TouchableOpacity>
    )
}

const s = StyleSheet.create({
    back: {
        marginLeft: 8,
        flexDirection: 'row'
    },
});

export default Back;