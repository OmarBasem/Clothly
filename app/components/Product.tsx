import React from 'react'
import {Text, TouchableOpacity, StyleSheet} from 'react-native'
import {useDispatch} from "react-redux";
import FastImage from "react-native-fast-image";
import {useNavigation} from '@react-navigation/native';
import {selectProduct} from "../reducers/products";

import {W} from '../utils'
import {product} from "../types";
import {HomeScreenNavigationProp} from "../navigators/types";


interface IProductProps {item: product, row: number, col: number}

// Product item on the HomeScreen
const Product = ({item, row, col} : IProductProps) => { // row and col are needed for testing purposes
    const navigation = useNavigation<HomeScreenNavigationProp>()
    const dispatch = useDispatch()
    const openProduct = () => {
        dispatch(selectProduct(item.id))
        navigation.navigate('Product')
    }
    return (
        <TouchableOpacity activeOpacity={1} style={s.container} onPress={() => openProduct()} testID={`item-${row}-${col}`}>
            <FastImage source={{uri: item.imageURL}} style={s.image} resizeMode='contain'/>
            <Text style={s.title} testID={`title-${row}-${col}`}>{item.title}</Text>
            <Text style={s.price}>AED {item.price}</Text>
        </TouchableOpacity>
    )
}

const s = StyleSheet.create({
    image: {
        width: W(40),
        height: W(60),
    },
    container: {
        width: W(45),
        padding: 8,
        borderWidth: 1,
        borderColor: 'lightgrey',
        marginBottom: 8,
        marginLeft: 12
    },
    title: {
        fontSize: 16
    },
    price: {
        fontSize: 16
    }
})

export default Product;