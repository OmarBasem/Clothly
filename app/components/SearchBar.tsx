import React, {useState} from 'react';
import {View, TextInput, StyleSheet} from 'react-native';
import IoIcon from "react-native-vector-icons/Ionicons";
import {theme} from "../styles";
import {W} from '../utils'
import {useDispatch} from "react-redux";
import {filterProducts} from "../reducers/products";

const SearchBar = () => {
    const [input, setInput] = useState('')
    const dispatch = useDispatch();
    const search = (input: string) => {
        setInput(input)
        dispatch(filterProducts(input))
    }
    return (
        <View style={s.container}>
            <TextInput style={s.input} selectionColor={theme.mainColor} placeholder='SEARCH'
                       value={input}
                       onChangeText={search} testID='SearchBar'/>
            <IoIcon name='ios-search' color='grey' size={20} style={s.search}/>
        </View>
    )
}

const s = StyleSheet.create({
    input: {
        width: W(96),
        height: 40,
        alignSelf: 'center',
        borderRadius: 8,
        borderWidth: 1,
        borderColor: 'grey',
        fontSize: 16,
        paddingLeft: 8,
        paddingRight: 32,
        marginBottom: 4,
    },
    container: {
        flexDirection: 'row',
        alignSelf: 'center',
    },
    cancel: {
        position: 'absolute',
        right: 4,
        top: 24
    },
    search: {
        position: 'absolute',
        zIndex: 1,
        right: 8,
        top: 8
    }
});

export default SearchBar;