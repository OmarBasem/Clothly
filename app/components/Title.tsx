import React from 'react';
import {Text, StyleSheet} from 'react-native';
import {theme} from "../styles";


// Screens Title component
const Title = ({text} : {text: string}) => {
    return (
        <Text numberOfLines={1} style={[s.title, {
            color: theme.mainColor,
            textAlign: 'center',
        }]}>{text}</Text>
    )
}

const s = StyleSheet.create({
    title: {
        fontWeight: '600',
        fontSize: 28,
        zIndex: 100,
        textAlign: 'left',
    }
});


export default Title;
