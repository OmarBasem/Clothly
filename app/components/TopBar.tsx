import React from 'react'
import {Text, View, StyleSheet, TouchableOpacity} from "react-native";
import {theme} from "../styles";
import {useDispatch, useSelector} from "react-redux";
import {useNavigation} from '@react-navigation/native';

import {changeCategory} from "../reducers/products";
import {HomeScreenNavigationProp} from "../navigators/types";
import {IApplicationState} from "../types";
import {fetchProductsAction} from "../actions";

const TopBar = () => {
    const categories = ['MEN', 'WOMEN', 'KIDS', 'FTW', 'ACCESSORIES']
    const navigation = useNavigation<HomeScreenNavigationProp>()
    const {products, currentCategory} = useSelector((state: IApplicationState) => ({
        currentCategory: state.products.currentCategory,
        products: state.products.categories
    }));
    const dispatch = useDispatch();
    const fetchCategory = async (category: string) => {
        await navigation.navigate('Home')
        dispatch(changeCategory(category))
        if (!products[category])
            dispatch(fetchProductsAction(category))
    }
    return (
        <View style={s.topBar} testID='TopBar'>
            {categories.map(category => {
                const selected = category === currentCategory;
                return (
                    <TouchableOpacity onPress={() => fetchCategory(category)} key={category} testID={category}>
                        <Text style={[s.barText, selected ? {color: theme.mainColor, textDecorationLine: 'underline'} : {}]}>{category}</Text>
                    </TouchableOpacity>
                )

            })}
        </View>
    )
}

const s = StyleSheet.create({
    topBar: {
        width: '96%',
        flexDirection: 'row',
        justifyContent: 'space-between',
        paddingTop: 8,
        paddingBottom: 8,
        alignSelf: 'center'
    },
    barText: {
        color: 'grey',
        fontWeight: 'bold',
        fontSize: 16
    }
})

export default TopBar;