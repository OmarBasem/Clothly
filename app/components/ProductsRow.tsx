import React from 'react';
import {View} from "react-native";
import Product from './Product'
import {W} from '../utils'
import {product} from "../types";


interface IProductsRowProps {products: Array<product>, row: number}

// function to render each product row
const ProductsRow = ({products, row} : IProductsRowProps) => {
    return (
        <View style={{flexDirection: 'row', width: W(100)}}>
            {products.map((product, i) => product &&
                <Product item={product} row={row} col={i} key={`product-${product.id.toString()}`}/>)}
        </View>
    )
}

export default ProductsRow;