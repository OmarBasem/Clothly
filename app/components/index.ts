import Title from "./Title";
import Back from './Back';
import TopBar from "./TopBar";
import SearchBar from "./SearchBar";
import Product from "./Product";
import ProductsRow from "./ProductsRow";


export {
    Title,
    Back,
    SearchBar,
    TopBar,
    Product,
    ProductsRow
}