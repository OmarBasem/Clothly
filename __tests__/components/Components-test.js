import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store'

import {Back, Title, TopBar, SearchBar, Product, ProductsRow} from "../../app/components";

const testProduct = require('../data/testProduct.json')

describe('Testing components rendering', () => {
    test('Back button renders correctly', () => {
        const tree = renderer.create(<Back/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Title renders correctly', () => {
        const tree = renderer.create(<Title/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('TopBar renders correctly', () => {
        const mockStore = configureStore()
        const store = mockStore({products: {}})
        const tree = renderer.create(<Provider store={store}><TopBar/></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('SearchBar renders correctly', () => {
        const mockStore = configureStore()
        const store = mockStore()
        const tree = renderer.create(<Provider store={store}><SearchBar/></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Product renders correctly', () => {
        const mockStore = configureStore()
        const store = mockStore()
        const tree = renderer.create(<Provider store={store}><Product item={testProduct}/></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('ProductRow renders correctly', () => {
        const mockStore = configureStore()
        const store = mockStore()
        const tree = renderer.create(<Provider store={store}><ProductsRow
            products={[testProduct]}/></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
})

