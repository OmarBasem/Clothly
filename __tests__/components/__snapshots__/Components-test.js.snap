// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing components rendering Back button renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  hitSlop={
    Object {
      "right": 40,
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexDirection": "row",
      "marginLeft": 8,
      "opacity": 1,
    }
  }
  testID="back"
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#1C284F",
          "fontSize": 40,
        },
        undefined,
        Object {
          "fontFamily": "Ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`Testing components rendering Product renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "lightgrey",
      "borderWidth": 1,
      "marginBottom": 8,
      "marginLeft": 12,
      "opacity": 1,
      "padding": 8,
      "width": 337.5,
    }
  }
  testID="item-undefined-undefined"
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "height": 450,
          "width": 300,
        },
      ]
    }
  >
    <FastImageView
      resizeMode="contain"
      source={
        Object {
          "uri": "https://www.juniorcouture.ae/dw/image/v2/BGHV_PRD/on/demandware.static/-/Sites-master-catalog/default/dwdc8636f2/images/10014476/KEN-K25100_1.jpg?sw=800&sh=800",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <Text
    style={
      Object {
        "fontSize": 16,
      }
    }
    testID="title-undefined-undefined"
  >
    Tiger Medium Red
  </Text>
  <Text
    style={
      Object {
        "fontSize": 16,
      }
    }
  >
    AED 
    250
  </Text>
</View>
`;

exports[`Testing components rendering ProductRow renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "width": 750,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderColor": "lightgrey",
        "borderWidth": 1,
        "marginBottom": 8,
        "marginLeft": 12,
        "opacity": 1,
        "padding": 8,
        "width": 337.5,
      }
    }
    testID="item-undefined-0"
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "height": 450,
            "width": 300,
          },
        ]
      }
    >
      <FastImageView
        resizeMode="contain"
        source={
          Object {
            "uri": "https://www.juniorcouture.ae/dw/image/v2/BGHV_PRD/on/demandware.static/-/Sites-master-catalog/default/dwdc8636f2/images/10014476/KEN-K25100_1.jpg?sw=800&sh=800",
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
      testID="title-undefined-0"
    >
      Tiger Medium Red
    </Text>
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
    >
      AED 
      250
    </Text>
  </View>
</View>
`;

exports[`Testing components rendering SearchBar renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="SEARCH"
    selectionColor="#1C284F"
    style={
      Object {
        "alignSelf": "center",
        "borderColor": "grey",
        "borderRadius": 8,
        "borderWidth": 1,
        "fontSize": 16,
        "height": 40,
        "marginBottom": 4,
        "paddingLeft": 8,
        "paddingRight": 32,
        "width": 720,
      }
    }
    testID="SearchBar"
    value=""
  />
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "grey",
          "fontSize": 20,
        },
        Object {
          "position": "absolute",
          "right": 8,
          "top": 8,
          "zIndex": 1,
        },
        Object {
          "fontFamily": "Ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`Testing components rendering Title renders correctly 1`] = `
<Text
  numberOfLines={1}
  style={
    Array [
      Object {
        "fontSize": 28,
        "fontWeight": "600",
        "textAlign": "left",
        "zIndex": 100,
      },
      Object {
        "color": "#1C284F",
        "textAlign": "center",
      },
    ]
  }
/>
`;

exports[`Testing components rendering TopBar renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingBottom": 8,
      "paddingTop": 8,
      "width": "96%",
    }
  }
  testID="TopBar"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="MEN"
  >
    <Text
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {},
        ]
      }
    >
      MEN
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="WOMEN"
  >
    <Text
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {},
        ]
      }
    >
      WOMEN
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="KIDS"
  >
    <Text
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {},
        ]
      }
    >
      KIDS
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="FTW"
  >
    <Text
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {},
        ]
      }
    >
      FTW
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="ACCESSORIES"
  >
    <Text
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 16,
            "fontWeight": "bold",
          },
          Object {},
        ]
      }
    >
      ACCESSORIES
    </Text>
  </View>
</View>
`;
